nomad/node_endpoint.go

```diff
@@ -435,6 +444,9 @@ func (n *Node) deregister(args *structs.NodeBatchDeregisterRequest,
 
 // UpdateStatus is used to update the status of a client node
 func (n *Node) UpdateStatus(args *structs.NodeUpdateStatusRequest, reply *structs.NodeUpdateResponse) error {
+	identity, authErr := n.srv.Authenticate(n.ctx, args.AuthToken)
+	args.SetIdentity(identity)
```

Comment:
Instead of having to remember to call `args.SetIdentity` after every call to `Authenticate`, what if made `Server.Authenticate` (or a new wrapper around `Authenticate`) take an interface that could both retrieve the auth token and set the identity?

This is just a really non-idiomatic pattern to expect devs to get right in every RPC: both understanding when to call SetIdentity (always?) as well as knowing where exactly to check `authErr` (*after* forwarding?).

Reply:
An alternative might be to actually update our ancient `Server.forward` method (or make a new method to migrate RPCs piecemeal) to just encapsulate *all* of this. This would reduce the per-RPC boilerplate as well as only have to write the slightly unusual "check error *after* a subsequent operation" authErr handling.

Reply:
We could definitely call `SetIdentity` inside `Authenticate`, as we always want it. Should be easy enough to do.

But although I love the idea of wrapping it into `Server.forward` in principle, there's a bunch of leader-only RPCs like `Eval.Dequeue` that don't forward because we _always_ send them to the leader. This is why I was exploring that RPC middleware notion previously (but the type signatures are _terrible_ :grin: )

Reply:
Ok, I've moved the `SetIdentity` into `Authenticate`, which also cleans up `Authenticate`'s signature so that if we eventually get it to produce ACLs for claims we don't have to change it again everywhere

Reply:
I've also updated the semgrep rule so that if we're using `Authenticate` we expect to return the error at some point. That should minimize the risk of misuse.